expression-language.md

```diff
@@ -0,0 +1,369 @@
+operators.
+
+```ebnf
+not-operator = "NOT"
+unary-logical-operator = not-operator
+binary-logic-operator = "AND" | "OR" | "XOR"
+
+unary-numeric-operator = "-"
+binary-numeric-comparison-operator = "=" | ">=" | "<=" | "<" | ">"
+binary-numeric-arithmetic-operator = "+" | "-" | "*" | "/" | "%"
+
+like-operator = "LIKE"
+exists-operator = "EXISTS"
+in-operator = "IN"
+
+unary-operation ::= (unary-numeric-operator | unary-logical-operator) expression
+
+binary-operation ::= expression (binary-numeric-comparison-operator | binary-numeric-arithmetic-operator | binary-logic-operator) expression
+
+like-operation ::= expression not-operator? like-operator string-literal
+
+exists-operation ::= exists-operator value-identifier
+
+set-expression ::= "(" expression ("," expression)* ")"
+in-operation ::= expression not-operator? in-operator set-expression
+```
+
+Note: the `set-expression` supports only compile-time set definition, hence the language doesn't include a _set_ type
+that can be manipulated at runtime.
+
+### 2.4. Functions invocation
+
+CESQL supports N arity function invocation:
+
+```ebnf
+char ::= [A-Z] | [a-z]
+parameter ::= expression
+
+function-identifier ::= char ( "_" | char )*
+
+parameter-list ::= parameter ("," parameter)*
+function-invocation ::= function-identifier "(" parameter-list? ")"
+```
+
+## 3. Language semantics
+
+### 3.1. Type System
+
+The type system contains 3 types:
+
+- _String_: Sequence of allowable Unicode characters
+- _Integer_: A signed 32-bit integer
+- _Boolean_: A boolean value of "true" or "false"
+
+The [types defined in the CloudEvents specification][ce-type-system] URI, URI Reference and String are represented as
+_String_.
+
+### 3.2. CloudEvent context identifiers and types
+
+Each CloudEvent context attribute and extension MUST be addressable from an expression using its identifier, as defined
+by the spec. For example, using `id` in an expression will address to the CloudEvent [id attribute][ce-id-attribute].
+
+Unless otherwise specified, every attribute and extension MUST be represented by the _String_ type. Through explicit
+casting, the user can convert the addressed value instances to _Integer_ and _Boolean_.
+
+When addressed an attribute not included in the input event, an empty _String_ MUST be assumed as value
+
+### 3.3. Errors
+
+Because every operator and function is total, an expression evalution flow is defined statically and cannot be modified
+by expected or unexpected errors. Nevertheless CESQL includes the concept of errors: when an expression is evaluated, in
+case an error arise, the evaluator collects a list of errors, referred in this spec as _error list_, which is then
+returned together with the evaluated value of the CESQL expression.
+
+Whenever possible, some errors checks SHOULD be done at compile time by the expression evaluator, in order to prevent
```

Comment:
What are some examples of the compile time expression evaluation? If the SQL parser is going to be part of the sdk so people can parse the events they receive for example, then we could have some fluent api with some type safety eg. catch an invalid cast at compile time.

Reply:
> catch an invalid cast at compile time.

That's an example: you're trying to convert `time` to _Boolean_ with `BOOL(time)`. This can fail at compile time, since it never makes sense to convert `time` to _Boolean_